lightning/src/ln/onion_utils.rs

```diff
@@ -376,229 +391,269 @@ pub(super) fn build_first_hop_failure_packet(shared_secret: &[u8], failure_type:
-						}
-					} else {
-						// Useless packet that we can't use but it passed HMAC, so it
-						// definitely came from the peer in question
-						let network_update = Some(NetworkUpdate::NodeFailure {
-							node_id: route_hop.pubkey,
-							is_permanent: true,
-						});
-						let short_channel_id = Some(route_hop.short_channel_id);
-						res = Some((network_update, short_channel_id, !is_from_final_node));
-					}
+		(path, session_priv, first_hop_htlc_msat)
+	} else { unreachable!() };
+	let mut res = None;
+	let mut htlc_msat = *first_hop_htlc_msat;
+	let mut error_code_ret = None;
+	let mut error_packet_ret = None;
+	let mut is_from_final_node = false;
+
+	// Handle packed channel/node updates for passing back for the route handler
+	construct_onion_keys_callback(secp_ctx, &path, session_priv,
+		|shared_secret, _, _, route_hop_opt, route_hop_idx|
+	{
+		if res.is_some() { return; }
+
+		let ammag = gen_ammag_from_shared_secret(shared_secret.as_ref());
+
+		let mut decryption_tmp = Vec::with_capacity(packet_decrypted.len());
+		decryption_tmp.resize(packet_decrypted.len(), 0);
+		let mut chacha = ChaCha20::new(&ammag, &[0u8; 8]);
+		chacha.process(&packet_decrypted, &mut decryption_tmp[..]);
+		packet_decrypted = decryption_tmp;
+
+		let err_packet = match msgs::DecodedOnionErrorPacket::read(&mut Cursor::new(&packet_decrypted)) {
+			Ok(p) => p,
+			Err(_) => return
+		};
+		let um = gen_um_from_shared_secret(shared_secret.as_ref());
+		let mut hmac = HmacEngine::<Sha256>::new(&um);
+		hmac.input(&err_packet.encode()[32..]);
+
+		// The failing hop includes either the inbound channel to the recipient or the outbound channel
+		// from the current hop (i.e., the next hop's inbound channel).
+		let num_blinded_hops = path.blinded_tail.as_ref().map_or(0, |t| t.hops.len());
+		is_from_final_node = route_hop_idx + 1 == path.hops.len() &&
+			(path.blinded_tail.is_none() || num_blinded_hops == 1);
+
+		if !fixed_time_eq(&Hmac::from_engine(hmac).into_inner(), &err_packet.hmac) { return }
+		let error_code_slice = match err_packet.failuremsg.get(0..2) {
+			Some(s) => s,
+			None if route_hop_opt.is_some() => {
+				// Useless packet that we can't use but it passed HMAC, so it definitely came from the peer
+				// in question
+				let network_update = Some(NetworkUpdate::NodeFailure {
+					node_id: route_hop_opt.as_ref().unwrap().pubkey,
+					is_permanent: true,
+				});
+				let short_channel_id = Some(route_hop_opt.as_ref().unwrap().short_channel_id);
+				res = Some((network_update, short_channel_id, !is_from_final_node));
+				return
+			},
+			None => return, // Useless packet from a blinded path hop
+		};
+
+		let error_code = u16::from_be_bytes(error_code_slice.try_into().expect("len is 2"));
+		error_code_ret = Some(error_code);
+		error_packet_ret = Some(err_packet.failuremsg[2..].to_vec());
+
+		let route_hop = match route_hop_opt {
+			Some(hop) => hop,
+			None => {
+				res = Some((None, None, true));
```

Comment:
For all these lines where we set `res = Some((None, None, true));` when the error is from a blinded hop, do we set them to be retryable just because we don't have a way of knowing otherwise?

Reply:
Yeah, if we return `payment_retryable = false` here then we mark the payment as abandoned in `outbound_payment`, which would be bad since it would prevent further payment retries. A payment is only not retryable if the recipient explicitly rejects it, which we have no way of knowing with blinded path errors.